core_lib/src/managers/selectionmanager.cpp

```diff
@@ -0,0 +1,370 @@
+#include "selectionmanager.h"
+#include "editor.h"
+
+#include "layerbitmap.h"
+#include "vectorimage.h"
+#include "bitmapimage.h"
+
+#include "layervector.h"
+
+//#ifdef QT_DEBUG
+#include <QDebug>
+//#endif
+
+
+SelectionManager::SelectionManager(Editor* editor) : BaseManager(editor),
+    mySelection(QRectF()), myTempTransformedSelection(QRectF()), myTransformedSelection(QRectF()),
+    mSomethingSelected(false), mLastSelectionPolygonF(QPolygonF()), mCurrentSelectionPolygonF(QPolygonF()),
+    myRotatedAngle(0), mOffset(QPointF()), mMoveMode(MoveMode::NONE)
+{
+}
+
+SelectionManager::~SelectionManager()
+{
+}
+
+bool SelectionManager::init()
+{
+    return true;
+}
+
+Status SelectionManager::load(Object*)
+{
+    return Status::OK;
+}
+
+Status SelectionManager::save(Object*)
+{
+    return Status::OK;
+}
+
+void SelectionManager::workingLayerChanged(Layer *)
+{
+}
+
+void SelectionManager::resetSelectionTransformProperties()
+{
+    mOffset = QPoint(0, 0);
+    myRotatedAngle = 0;
+    mSelectionTransform.reset();
+}
+
+
+void SelectionManager::updatePolygons()
+{
+    mCurrentSelectionPolygonF = myTempTransformedSelection;
+    mLastSelectionPolygonF = myTransformedSelection;
+}
+
+void SelectionManager::resetSelectionTransform()
+{
+    mSelectionTransform.reset();
+}
+
+bool SelectionManager::isOutsideSelectionArea(QPointF point)
+{
+    return (!myTransformedSelection.contains(point)
+            && validateMoveMode(point) == MoveMode::NONE);
+}
+
+void SelectionManager::deleteSelection()
+{
+    emit needDeleteSelection();
+}
+
+void SelectionManager::clearCurves()
+{
+    mClosestCurves.clear();
+}
+
+void SelectionManager::clearVertices()
+{
+    mClosestVertices.clear();
+}
+
+float SelectionManager::selectionTolerance() const
+{
+    return mSelectionTolerance * editor()->viewScaleInversed();
+}
+
+MoveMode SelectionManager::validateMoveMode(QPointF pos)
+{
+    return moveModeForAnchorInRange(pos);
+}
+
+MoveMode SelectionManager::moveModeForAnchorInRange(QPointF lastPos)
+{
+    const double marginInPixels = 15;
+    const double scale = editor()->viewScaleInversed();
+    const double scaledMargin = qAbs(marginInPixels * scale);
```

Comment:
Then we have to make sure that selectionTolerance is updated at the time of using it, otherwise the value will be wrong.

Reply:
But it calculates the scaling when called, in almost the exact same way as here. The only difference (aside from the difference between this hardcoded value and the one for mSelectionTolerance) is that we take the absolute value here but not in selectionTolerance. Maybe we should be taking the absolute value in selectionTolerance too in case viewScaleInversed is ever negative? Couldn't that happen if the view is flipped?

Reply:
Hmm ah I guess that's a valid point, alright I'll refactor so that it uses selectionTolerance and tolerance will be normalised too.